Miki.Discord.Common/Models/IDiscordGuild.cs

```diff
@@ -37,25 +37,61 @@ public interface IDiscordGuild : ISnowflake
 
 		GuildPermission Permissions { get; }
 
+        /// <summary>
+        /// Bans a discord user from the discord guild.
+        /// </summary>
+        /// <param name="user">The user in question.</param>
+        /// <param name="pruneDays">Amount of days of messages sent to be pruned</param>
+        /// <param name="reason">Reason of the ban</param>
 		Task AddBanAsync(IDiscordGuildUser user, int pruneDays = 1, string reason = null);
 
+        /// <summary>
+        /// Creates a new role. Requires <see cref="GuildPermission.ManageRoles"/> permission.
+        /// </summary>
+        /// <param name="roleArgs">Role properties</param>
 		Task<IDiscordRole> CreateRoleAsync(CreateRoleArgs roleArgs);
 
+        /// <summary>
+        /// Gets the system channel.
+        /// </summary>
 		Task<IDiscordChannel> GetDefaultChannelAsync();
 
-		Task<GuildPermission> GetPermissionsAsync(IDiscordGuildUser user);
+        /// <summary>
+        /// Get permissions that are attached the <see cref="IDiscordGuildUser"/> based on the guild's permissions and the user's roles.
+        /// </summary>
+        Task<GuildPermission> GetPermissionsAsync(IDiscordGuildUser user);
 
+        /// <summary>
+        /// Gets the <see cref="IDiscordGuildUser"/> of the current discord server owner.
+        /// </summary>
 		Task<IDiscordGuildUser> GetOwnerAsync();
 
-		Task<IDiscordGuildChannel> GetChannelAsync(ulong id);
+        /// <summary>
+        /// Gets a <see cref="IDiscordGuildChannel"/> where its snowflake matches {id}. Will return null if non-existant.
+        /// </summary>
+        /// <param name="id">Snowflake of the channel.</param>
+        Task<IDiscordGuildChannel> GetChannelAsync(ulong id);
 
-		Task<IEnumerable<IDiscordGuildChannel>> GetChannelsAsync();
+        /// <summary>
+        /// Returns all channels in a guild.
+        /// </summary>
+        Task<IEnumerable<IDiscordGuildChannel>> GetChannelsAsync();
 
+        /// <summary>
+        /// Gets a <see cref="IDiscordRole"/> where its snowflake matches {id}. Will return null if non-existant.
+        /// </summary>
+        /// <param name="id">Snowflake of the role.</param>
 		Task<IDiscordRole> GetRoleAsync(ulong id);
 
+        /// <summary>
+        /// Returns all roles in a guild.
+        /// </summary>
 		Task<IEnumerable<IDiscordRole>> GetRolesAsync();
 
-		Task<IDiscordGuildUser[]> GetMembersAsync();
+        /// <summary>
+        /// Returns all members in a guild.
+        /// </summary>
+		Task<IEnumerable<IDiscordGuildUser>> GetMembersAsync();
```

Comment:
Are we going to use `IEnumerable` or `IReadOnlyList` as return types for lists?

Personally, I'm leaning towards `IReadOnlyList` since the data is already fetched from Discord / the cache so it's most likely already an array or a list, which can be directly casted towards a `IReadOnlyList`. With `IReadOnlyList` we can get the length of the list without creating an `IEnumerator` (which is created by `enumerable.Count()`).

Maybe in the future when .NET Standard 2.1 is released we could change these methods to return a `IAsyncEnumerable`.

Reply:
I'm not sure. Once AsyncEnumerable comes out soon, adopting that seems the easiest answer. and with that, the best solution for now. It would make sense to adopt Enumerable to make sure nothing else changes for the time being.

Reply:
Alright, `IEnumerable` it is.